Make use of queues and add blinking platforms

Jontpan opened the pull request:
When testing, use basicPlatformScheme(30, 40, 5, 15, 10);

#96

Jontpan committed:
Change logic of platformspawner

Jontpan committed:
Add blinking platforms

This adds blinking platforms to the game. When jumped on, the next three platforms
will blink away and then reappear when jumped on.

There are some known bugs. If the blinkplatforms spawn too close to eachother
they will blink away eachothers next three. For know I'm happy to just make that risk
a low risk, but this needs to be considered.

Jontpan committed:
Add correct script

Jontpan committed:
Reset to normal basicplatformscheme

AxlLind commented on 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs:
```diff
@@ -0,0 +1,34 @@
+            rb.velocity = new Vector2(rb.velocity.x, jumpSpeed);
+            if (GameControllerScript.soundOn)
+            {
+                audioSource.Play();
```
This appears to give a null pointer exception when you jump on the platform

Jontpan replied:
Also something I believe I fixed in some of the later commits.

AxlLind commented on 2Dgame/Assets/Scripts/Platform Scripts/BlinkerScript.cs:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    /**
+     * Enables the renderer when jumped on
```
This doesn't seem to happen? Is the idea for them to be shown again when you jump on them again? When I've played they stay hidden.

Jontpan replied:
Check to see if the blinkerprefab has the correct script attached, I thought i fixed it in the add correct script commit. I may have unfixed it when i reset the basicplatformscheme....

Jontpan replied:
Don't seem to have this problem, you probably don't have the latest commits on this branch

AxlLind commented on 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs:
```diff
@@ -40,83 +45,81 @@ public class PlatformSpawnerScript : MonoBehaviour {
         {
-            return basicPlatformScheme();
+            basicPlatformScheme();
+            return;
```
These return statements are now redundant

Jontpan replied:
You are correct. For a minute there I thought the bug I had was that it fell through the if statements.

AxlLind reviewed (commented):
Looks good! Some small things I've left comments on but there is one major thing:

When there are two blinker platforms on the screen when you jump on the first one the next blinker's platforms get hidden as well.

Also if you've already jumped on a blinker (purple one) and then an other appears the hidden ones start blinking when jumping on the new blinker.

Jontpan commented:
Yep, I outlined these problems in the commit message. While I have some idea of how to fix it, I hoped that it would never occur... :P

Jontpan committed:
fix redundancy and a bug

AxlLind commented:
My bad, didn't read the commit message. Found a problem with the prefabqueue implementation. It keeps adding stuff to the queue even when there are stuff in there. This would be ok if we only added one prefab and took one out every time but we add more than one and only take out one.

This means the queue increases in length quickly. With some quick testing it was up to 50 elements at y = 200.

Jontpan commented:
I am aware of this but figured it wouldn't be too bad. But we could honestly just clear the queue at every difficulty increase.

AxlLind commented:
Cant we just add an if statement around the choosePlatform()?

``` C#
if (prefabQueue.Count == 0) 
{
    choosePlatform();
}
```

That way we don't add stuff to the queue unless we need to

Jontpan commented:
Lot better.

Jontpan committed:
Fix blinking bug and restructure scripts

Blinkingplatform now only grabs the next three blinkers.
Also keeps the prefabqueue from getting larger than 3.

Reference to #109, #96

Jontpan committed:
change to correct basicPlatformScheme

AxlLind commented on 2Dgame/Assets/Scripts/Platform Scripts/OnJumpScript.cs:
```diff
@@ -0,0 +1,44 @@
+
+    protected new void Awake()
+    {
+        notJumpedOn = false;
```
This needs to be set to true.

AxlLind commented on 2Dgame/Assets/Scripts/PlatformSpawnerScript.cs:
```diff
@@ -40,83 +49,75 @@ public class PlatformSpawnerScript : MonoBehaviour {
         else if (this.transform.position.y < 1500)
         {
-            return PlatformScheme(45, 40, 5, 10); // More hard stuff
+            PlatformScheme(30, 40, 5, 15, 10); // More hard stuff, first blink
```
I feel like blinkers could happen earlier maybe

Jontpan replied:
we should probably decide together exactly how the progression should be, maybe when we finalized the game for release. I'll leave it right now.

AxlLind commented on 2Dgame/Assets/Scripts/Platform Scripts/BlinkingPlatformScript.cs:
```diff
@@ -0,0 +1,56 @@
+    protected override void OnJump()
+    {
+        GameObject[] blinkers = GameObject.FindGameObjectsWithTag("Blinker");
+        StartCoroutine(gcs.PauseForSeconds(2));
```
Maybe this should be slightly shorter? I feel like it's fine for the flips but for the blinkers I think it should maybe be shorter.

Jontpan replied:
Tested, 1.2 seconds feels good.

AxlLind commented on 2Dgame/2Dgame.sln:
```diff
@@ -1,29 +1,26 @@
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{3A334A53-003A-5D3C-1187-9E32BAE4F6C3}.Release|Any CPU.Build.0 = Release|Any CPU
+		{DE6F5ADE-BF29-988E-0D90-29C150F7701D}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Shouldn't this be "DE6F5ADE-BF29-988E-0D90-29C150F7701G"?

Jontpan replied:
o shit haha. how long has this bug been here lol

AxlLind commented on 2Dgame/Assets/Scripts/Scene Controllers/GameControllerScript.cs:
```diff
@@ -80,9 +81,18 @@ public class GameControllerScript : MonoBehaviour {
 	}
 
-	// Button press-methods
+    public IEnumerator PauseForSeconds(float seconds)
```
This is a really nice function!

AxlLind reviewed (approved):
Looks really good! I really like the blinking effect 👍  The code looks much better now with the queue and with the OnJump() change. Nice work!

Just fix the NotJumpedOn to true instead of false and we're good!

Jontpan committed:
Fix bug and shorten pausetime
